opentelemetry-api/src/common.rs

```diff
@@ -451,18 +451,40 @@ pub struct InstrumentationLibrary {
     ///
     /// [Schema url]: https://github.com/open-telemetry/opentelemetry-specification/blob/v1.9.0/specification/schemas/overview.md#schema-url
     pub schema_url: Option<Cow<'static, str>>,
+
+    /// Specifies the instrumentation scope attributes to associate with emitted telemetry.
+    pub attributes: Option<Vec<KeyValue>>,
+
+}
+
+// Uniqueness for InstrumentationLibrary/InstrumentationScope does not depend on attributes
+impl Eq for InstrumentationLibrary {}
+
+impl PartialEq for InstrumentationLibrary {
+    fn eq(& self, other: &Self) -> bool {
+        self.name == other.name && self.version == other.version && self.schema_url == other.schema_url
+    }
+}
+
+impl hash::Hash for InstrumentationLibrary {
+    fn hash<H: hash::Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.version.hash(state);
+        self.schema_url.hash(state);
+    }
 }
 
 impl InstrumentationLibrary {
     /// Create an new instrumentation library.
-    pub fn new<T>(name: T, version: Option<T>, schema_url: Option<T>) -> InstrumentationLibrary
+    pub fn new<T>(name: T, version: Option<T>, schema_url: Option<T>, attributes: Option<Vec<KeyValue>>) -> InstrumentationLibrary
```

Comment:
With the idea to move tracing to 1.0 in mind, I am a little worry how we handle new parameters here. Since spec doesn't consider adding new parameters a breaking change we should probably make it not a breaking change in our API.

Reply:
Yeah that's a good point. Also with potential other new additions in the future I wonder if a builder makes the most sense where libraries are generally specified. e.g. 

```rust
let tracer = TracerProvider::tracer_builder("tracer_name")
    .with_version("1.0.0")
    .with_attributes([kv])
    .build();
```

Reply:
Or something along those lines to prevent future API changes.

Reply:
`Option<Vec<KeyValue>>` works here but may not be the most ergonomic. Maybe `Option<V> where V: Into<Vec<KeyValue>>` so that user can pass a `Vec`, a `&[KeyValue]` or other struct implemented `Into<Vec<T>>` trait

Reply:
@TommyCpp 

I changed the function signature for `versioned_tracer` to look like this:

```rust
fn versioned_tracer(
        &self,
        name: impl Into<Cow<'static, str>>,
        version: Option<&'static str>,
        schema_url: Option<&'static str>,
        attributes: Option<impl Into<Vec<opentelemetry_api::KeyValue>>>,
    )
```
to try to allow any struct that implements `Into<Vec<T>>`. This forces me to pass in explicit typing whenever I use `versioned_tracer()` like below:

```rust
fn tracer(&self, name: impl Into<Cow<'static, str>>) -> Self::Tracer {
        self.versioned_tracer(name, None, None, None::<Vec<KeyValue>>)
    }
```
Is this the correct typing or is it unavoidable to always have to specify type information, even if the value is `None`?

Reply:
> Just to clarify, did you mean it would be desirable to have our api that uses InstrumentationLibrary (versioned_tracer()) accept different types of input for the attributes field

Yes that's my intention. It's similar to other parameters. For `Option<T>` struct in Rust has to know the actual type of `T` during compile time. 

I think at his point we can just use `Option<Vec<KeyValue>>` for now.  A better solution will be using builder pattern and get rid of the `Option`. Then we can ask user just provide any type that can convert to `Vec<KeyValue>`

Reply:
Is it possible to create "overloads" of `versioned_tracer`?

fn versioned_tracer(..)
{
versioned_tracer(..,None::<Vec<KeyValue>>)
}

fn versioned_tracer(.., attributes: Option<Vec<opentelemetry_api::KeyValue>>)

Reply:
Just learned Rust has no overloading, so ignore my above comment.